router/core/src/main/scala/io/buoyant/router/RoutingFactory.scala

```diff
@@ -3,22 +3,46 @@ package io.buoyant.router
 import com.twitter.finagle.{param => _, _}
 import com.twitter.finagle.buoyant.{Dst, DstBindingFactory}
 import com.twitter.finagle.tracing.Trace
-import com.twitter.util.{Future, Time}
+import com.twitter.util.{Throw, Return, Future, Time}
 import scala.util.control.NoStackTrace
 
 object RoutingFactory {
   val role = Stack.Role("RoutingFactory")
   val description = "Performs per-request name binding"
 
-  case class UnknownDst[Req](request: Req, cause: Throwable)
-    extends Exception(s"Unknown destination: $request / ${cause.getMessage}", cause)
-    with NoStackTrace
-
   /**
    * An Identifier determines a [[com.twitter.finagle.buoyant.Dst
    * destination]] for `Req`-typed requests.
    */
-  type Identifier[Req] = Req => Future[(Dst, Req)]
+  type Identifier[Req] = Req => Future[RequestIdentification[Req]]
+
+  /** The result of attempting to identify a request. */
+  sealed trait RequestIdentification[Req]
+  /**
+   * This indicates that a destination was successfully assigned.  The attached
+   * request should be sent to the destination.
+   * @param dst The destination of the request.
+   * @param request The request to send to the destination.  This allows
+   *                identifiers to effectively mutate requests that they
+   *                identify.
+   */
+  class IdentifiedRequest[Req](val dst: Dst, val request: Req) extends RequestIdentification[Req]
+
+  object IdentifiedRequest {
+    def unapply[Req](identified: IdentifiedRequest[Req]): Option[(Dst, Req)] =
+      Some(identified.dst, identified.request)
+  }
+
+  /**
+   * This indicates that the identifier could not assign a destination to the
+   * request.
+   */
+  class UnidentifiedRequest[Req](val reason: String) extends RequestIdentification[Req]
```

Comment:
Another thought:

since this ADT is under RoutingFactory, it will be cumbersome to write Java identifiers (they'll have to use $s or, we'll have to provide `RequestIdentifications` compat builders